Source/Scene/ImplicitTileCoordinates.js

```diff
@@ -25,7 +25,8 @@ import ImplicitSubdivisionScheme from "./ImplicitSubdivisionScheme.js";
  *
```

Comment:
@IanLilleyT and I had a whiteboard session to design a cleaner interface for this file. 

![image](https://user-images.githubusercontent.com/8422414/116926586-85e55280-ac28-11eb-8791-8ab35df10977.png)

The diagram was a bit of a mess by the end, so here are the key functions we need:

**Fundamental functions**:

All three of these functions work whether using global coordinates (relative to the root implicit tile) or local (relative to a subtree root)

* `ImplicitTileCoordinates.prototype.deriveDescendant(offsetCoordinates)` - computes the coordinates of a tile deeper in the tree by combining the current coordinates + an offset. It's roughly analogous to vector addition `position + offset` going _down_ the tree.
* `ImplicitTileCoordinates.prototype.deriveAnscestor(offsetLevels)` - computes the coordinates of the tile higher up in the tree by going up a number of levels. Since this is on a tree, there's only ever one possible path so only a level offset is needed. This is also analogous to vector addition `position + offset` except going _up_ the tree.
* `ImplicitTileCoordinates.getRelativeCoordinates(anscestor, descendant)` (not sure the best signature, or if it should be a static method or an instance method) - compute the coordinates of descendant relative to ancestor. This is analogous to vector subtraction `descendant - anscestor`

**Other Functions**:

* `ImplicitTileCoordinates.prototype.deriveSubtreeCoordinates()` - exists for convenience, built atop `deriveAnscestor` using `offsetLevels = tileLevel - subtreeLevel`.
* `ImplicitTileCoordinates.prototype.deriveParentSubtreeCoordinates()` - exists for convenience, built atop `deriveAncestor`. `offsetLevels = (tileLevel - subtreeLevel) + subtreeLevels`
* `ImplicitTileCoordinates.deriveChildCoordinates(childIndex)` - We still need this since this is optimized for the transcoder (document it as such)
* `deriveLocalChildSubtreeCoordinates()` - actually we didn't say if we still needed this, but you can build it on top of `getRelativeCoordinates()` in any case. Though if we do keep this function, see if we can find a shorter name...

Reply:
`deriveLocalChildSubtreeCoordinates` could be implemented as `deriveRelativeCoordinates(tile.deriveParentSubtreeCoordinates(), tile)` and `deriveLocalTileCoordinates` could be implemented as `deriveRelativeCoordinates(tile.deriveSubtreeCoordinates(), tile)`

If that's too much to expect out of someone using the API, or too error-prone, we could have a helper function called `deriveChildSubtreeCoordinates` and `deriveLocalTileCoordinates`?

Reply:
Since you mention static vs instance, all function that take two inputs should be consistent. I think the API will be cleaner if it's all instance methods and it's made clear what is what. What do you think?

Reply:
For the first point, how often will those functions be used?

Hm, as far as `deriveRelativeCoordinates`, I guess the usage would normally be `descendant.deriveRelativeCoordinates(anscestor)`? or is the other way around better?

Reply:
hmm, the only place it's used is in `ImplicitSubtree`
```
ImplicitSubtree.prototype.childSubtreeIsAvailableAtCoordinates = function (
  implicitCoordinates
) {
  var index = this.getChildSubtreeIndex(implicitCoordinates);
  return this.childSubtreeIsAvailableAtIndex(index);
};
```
could become
```
ImplicitSubtree.prototype.childSubtreeIsAvailableAtCoordinates = function (
  implicitCoordinates
) {
  var subtreeCoordinates = implicitCoordinates.deriveParentSubtreeCoordinates();
  var offsetCoordinates = subtreeCoordinates.deriveOffset(implicitCoordinates);
  var index = this.getChildSubtreeIndex(offsetCoordinates);
  return this.childSubtreeIsAvailableAtIndex(index);
};
```